pulsar-broker-auth-oidc/src/main/java/org/apache/pulsar/broker/authentication/oidc/AuthenticationProviderOpenID.java

```diff
@@ -0,0 +1,441 @@
+    }
+
+    /**
+     * Authenticate the parameterized JWT.
+     *
+     * @param token - a nonnull JWT to authenticate
+     * @return a fully authenticated JWT, or AuthenticationException if the JWT is proven to be invalid in any way
+     */
+    private CompletableFuture<DecodedJWT> authenticateToken(String token) {
+        if (token == null) {
+            incrementFailureMetric(AuthenticationExceptionCode.ERROR_DECODING_JWT);
+            return CompletableFuture.failedFuture(new AuthenticationException("JWT cannot be null"));
+        }
+        final DecodedJWT jwt;
+        try {
+            jwt = decodeJWT(token);
+        } catch (AuthenticationException e) {
+            incrementFailureMetric(AuthenticationExceptionCode.ERROR_DECODING_JWT);
+            return CompletableFuture.failedFuture(e);
+        }
+        return verifyIssuerAndGetJwk(jwt)
+                .thenCompose(jwk -> {
+                    try {
+                        if (!jwt.getAlgorithm().equals(jwk.getAlgorithm())) {
+                            incrementFailureMetric(AuthenticationExceptionCode.ALGORITHM_MISMATCH);
+                            return CompletableFuture.failedFuture(
+                                    new AuthenticationException("JWK's alg [" + jwk.getAlgorithm()
+                                            + "] does not match JWT's alg [" + jwt.getAlgorithm() + "]"));
+                        }
+                        // Verify the JWT signature
+                        // Throws exception if any verification check fails
+                        return CompletableFuture
+                                .completedFuture(verifyJWT(jwk.getPublicKey(), jwt.getAlgorithm(), jwt));
+                    } catch (InvalidPublicKeyException e) {
+                        incrementFailureMetric(AuthenticationExceptionCode.INVALID_PUBLIC_KEY);
+                        return CompletableFuture.failedFuture(
+                                new AuthenticationException("Invalid public key: " + e.getMessage()));
+                    } catch (AuthenticationException e) {
+                        return CompletableFuture.failedFuture(e);
+                    }
+                });
+    }
+
+    private CompletableFuture<Jwk> verifyIssuerAndGetJwk(DecodedJWT jwt) {
+        // Verify that the issuer claim is nonnull and allowed.
+        if (jwt.getIssuer() == null || !this.issuers.contains(jwt.getIssuer())) {
+            incrementFailureMetric(AuthenticationExceptionCode.UNSUPPORTED_ISSUER);
+            return CompletableFuture
+                    .failedFuture(new AuthenticationException("Issuer not allowed: " + jwt.getIssuer()));
+        }
+        // Retrieve the metadata: https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderMetadata
+        return openIDProviderMetadataCache.getOpenIDProviderMetadataForIssuer(jwt.getIssuer())
+                .thenCompose(metadata -> jwksCache.getJwk(metadata.getJwksUri(), jwt.getKeyId()));
```

Comment:
Note that the value of `jwt.getIssuer()` here would match the _actual_ value of the `issuer` field in the metadata document.  It would not match the _configured_ issuer URL; see the GKE example elsewhere in this review.

Reply:
To clarify, this potential discrepancy is only a problem if we remove the equality check in the `verifyIssuer` method, right?

Reply:
I agree but do think it reasonable to use the _discovered_ issuer URL as the expected value of the `iss` claim.

Reply:
> I agree but do think it reasonable to use the _discovered_ issuer URL as the expected value of the `iss` claim.

This is now the case when the `openIDKubernetesDiscoveryMode` is `DISCOVER_TRUSTED_ISSUER` or `DISCOVER_PUBLIC_KEYS`.

I provide more explanation in my recent commits and I'll add some comments at the end of this PR.